### ISSUE TITLE
Support for "meta" members in the object identifier

### ISSUE THREAD

**Issue by peter-plochan** [toedter/spring-hateoas-jsonapi#21]
We have a use case when we need to enhance [JSON:API Resource Identifier](https://jsonapi.org/format/#document-resource-identifier-objects) by some optional "meta" information within the top level document as well as within individual relationships and includes. We could use the `meta` object for this which is allowed by the JSON:API specification on the identifier level:
> A “resource identifier object” **MAY** also include a `meta` member, whose value is a [meta](https://jsonapi.org/format/#document-meta) object that contains non-standard meta-information.

The result would look like the following example:
```yaml
{
    "data": {
        "id": "1",
        "type": "entity",
        "meta": {
            "idMetaField": "value1"
        },
        "attributes": {
            "description": "Entity with meta"
        },
        "relationships": {
            "relatedEntities": {
                "data": [{
                    "id": "2",
                    "type": "relatedEntity",
                    "meta": {
                        "idMetaField": "value2"
                    }
                }, {
                    "id": "3",
                    "type": "relatedEntity"
                    # we don't need any "meta" info for this relationship
                }, {
                    "id": "4",
                    "type": "relatedEntity",
                    "meta": {
                        "idMetaField": "value4"
                    }
                }]
            }
        }
    },
    "included": [{
        "id": "2",
        "type": "relatedEntity",
        "meta": {
            "idMetaField": "value2"
        },
        "attributes": {
            "title": "Related Entity 2"
        }
    }, {
        "id": "3",
        "type": "relatedEntity",
        "attributes": {
            "title": "Related Entity 3"
        }
    }, {
        "id": "4",
        "type": "relatedEntity",
        "meta": {
            "idMetaField": "value4"
        },
        "attributes": {
            "title": "Related Entity 4"
        }
    }],
    "links": {
        "self": "http://localhost/test/entities/1?include=relatedEntities"
    }
}
```
However, I cannot find any support for such member within the library. 

Consider to add support for this. Or is there any workaround or better solution? Thanks

**Comment by toedter** [comment]
The `JsonApiModelBuilder` provides a method 

``` Java
public JsonApiModelBuilder relationship(String name, Map<String, Object> meta)
```

where you can add meta to a relationship.
To add meta on the top-level document or in any data resource or included resource you could also use the `JsonApiModelBuilder` and then use

``` Java
public JsonApiModelBuilder meta(String key, Object value) 
```

Does this solve your problem?

**Comment by toedter** [comment]
I fixed this in the current snapshot, here is an example:

``` Java
        Director director = new Director("3", "George Lucas");
        final RepresentationModel<?> directorModel =
                jsonApiModel()
                        .model(EntityModel.of(director))
                        .meta("director-meta", "director-meta-value")
                        .build();

        Map<String, Object> relationshipMeta = new HashMap<>();
        relationshipMeta.put("relationship-meta", "relationship-meta-value");

        Movie movie = new Movie("1", "Star Wars");
        final RepresentationModel<?> movieModel =
                jsonApiModel()
                        .model(movie)
                        .meta("movie-meta", "movie-meta-value")
                        .relationship("directors", director)
                        .relationship("directors", relationshipMeta )
                        .build();

        final RepresentationModel<?> jsonApiModel =
                jsonApiModel()
                        .model(movieModel)
                        .meta("top-level-meta", "top-level-meta-value")
                        .included(directorModel)
                        .build();

        final String movieJson = mapper.writeValueAsString(jsonApiModel);

```

will be serialized to

``` JSON
{
  "data": {
    "id": "1",
    "type": "movies",
    "attributes": {
      "title": "Star Wars"
    },
    "relationships": {
      "directors": {
        "data": {
          "id": "3",
          "type": "directors"
        },
        "meta": {
          "relationship-meta": "relationship-meta-value"
        }
      }
    },
    "meta": {
      "movie-meta": "movie-meta-value"
    }
  },
  "included": [
    {
      "id": "3",
      "type": "directors",
      "attributes": {
        "name": "George Lucas"
      },
      "meta": {
        "director-meta": "director-meta-value"
      }
    }
  ],
  "meta": {
    "top-level-meta": "top-level-meta-value"
  }
}

```

This will be in the next release.

**Comment by peter-plochan** [comment]
Thank you for a quick fix, but the fix you introduced solves my problem partially. I need to have these `meta` objects in individual "related resource relationships" as well. Example:
```yaml
...
    "relationships": {
      "directors": {
        "data": [{
          "id": "3",
          "type": "directors",
          "meta": {
               "director-meta": "director3-meta-value"
          }
        }, {
          "id": "4",
          "type": "directors",
          "meta": {
               "director-meta": "director4-meta-value"
          }
        }],
        "meta": {
          "relationship-meta": "relationship-meta-value"
        }
      }
    },
...
```
and the representation model could be defined like this
```java
        Map<String, Object> director3Meta = new HashMap<>();
        director3Meta.put("director-meta", "director3-meta-value");

        Map<String, Object> director4Meta = new HashMap<>();
        director3Meta.put("director-meta", "director4-meta-value");

        Movie movie = new Movie("1", "Star Wars");
        final RepresentationModel<?> movieModel =
                jsonApiModel()
                        .model(movie)
                        .meta("movie-meta", "movie-meta-value")
                        .relationship("directors", director3, director3Meta)
                        .relationship("directors", director4, director4Meta)
                        .relationship("directors", relationshipMeta )
                        .build();
```
I'm not sure if it makes sense to you, but I have a use case where the `meta` object holds important info for individual resources even in the relationships part. This can be used by additional linkage info (e.g. when `id` and `type` fields are not sufficient to correctly identify related object) or when we don't want to show `included` section, but still we need the meta info for related objects.

Another problem is the deserialization - do you plan to support `meta` objects there? Or is it already supported somehow?

**Comment by toedter** [comment]
Thx Peter, I'll take a look at it.

**Comment by toedter** [comment]
Noe everything should work, see https://toedter.github.io/spring-hateoas-jsonapi/snapshot/reference/#Nesting

Please give your example another try.

**Comment by toedter** [comment]
released in 0.12.0

**Comment by peter-plochan** [comment]
Works, well, thanks a much 👍
